payments/providers/cpu_ceepos.py

```diff
@@ -0,0 +1,361 @@
+            product = order_line.product
+            int_tax = int(product.tax_percentage)
+            assert (
+                int_tax == product.tax_percentage
+            )  # make sure the tax is a whole number
+            items.append(
+                {
+                    "Code": product.sku,
+                    "Amount": order_line.quantity,
+                    "Price": price_as_sub_units(
+                        product.get_price_for_reservation(reservation)
+                    ),
+                    "Description": product.name,
+                    "Taxcode": int_tax,
+                }
+            )
+            payload["Products"] = items
+
+    def payload_add_customer(self, payload, order):
+        """Attaches customer data to the payload"""
+        reservation = order.reservation
+        payload.update(
+            {
+                "Email": reservation.billing_email_address,
+                "FirstName": reservation.billing_first_name,
+                "LastName": reservation.billing_last_name,
+            }
+        )
+
+    def payload_add_return_and_notification_urls(self, payload):
+        """Attaches return and notification URLs to the payload"""
+        payload.update(
+            {
+                "ReturnAddress": self.get_success_url(),
+                "NotificationAddress": self.get_notify_url(),
+            }
+        )
+
+    def payload_add_checksum(self, payload):
+        """
+        Attaches the checksum to the payload.
+
+        The checksum is calculated from a string that consists of the values of
+        the parameters in the payload and the source system specific secret key.
+        """
+        checksum_calc_values = []
+        for key, value in payload.items():
+            if isinstance(value, list):
+                for item in value:
+                    checksum_calc_values.extend(item.values())
+            else:
+                checksum_calc_values.append(value)
+        payload["Hash"] = self.calculate_checksum(checksum_calc_values)
+
+    def validate_ceepos_message_payload(self, request=None, response=None):
+        """
+        Validates that message payload received from CeePos has not been tampered with.
+        The received message is either a GET or POST request (success/notify) or
+        a response to the initiate payment request.
+
+        If the checksum calculated is the same as the Hash parameter in the request/response,
+        the message has been received intact and directly from CeePos.
+        """
+
+        REQUEST_CHECKSUM_PARAMS = (
+            "Id",
+            "Status",
+            "Reference",
+            "PaymentMethod",
+            "PaymentSum",
+            "Timestamp",
+            "PaymentDescription",
+        )
+        RESPONSE_CHECKSUM_PARAMS = (
+            "Id",
+            "Status",
+            "Reference",
+            "Action",
+            "PaymentAddress",
+            "PaymentExpires",
+        )
```

Comment:
These dict could be module level globals instead. And since this validation doesn't use the self parameter at all, it could even make sense to put all validation code to its own module and call it via couple free functions.

Reply:
OK, didn't notice that it actually uses the secret, but it would still make sense even if you have to pass that secret in. And it actually makes it more explicit to the reader of the code that there is some secret that plays a part in this validation.

Reply:
Moved them to the module global level. Also the secret is now passed in to the validation functions (see the other comment below https://github.com/andersinno/respa/pull/27#discussion_r558196487).